controllers/machinedeployment_sync_test.go

```diff
@@ -397,3 +398,109 @@ func TestScaleMachineSet(t *testing.T) {
 		})
 	}
 }
+
+func newDeploymentStatus(replicas, updatedReplicas, availableReplicas int32) clusterv1.MachineDeploymentStatus {
+	return clusterv1.MachineDeploymentStatus{
+		Replicas:          replicas,
+		UpdatedReplicas:   updatedReplicas,
+		AvailableReplicas: availableReplicas,
+	}
+}
+
+func currentDeployment(pds *int32, replicas, statusReplicas, updatedReplicas, availableReplicas int32, conditions clusterv1.Conditions) *clusterv1.MachineDeployment {
+	d := &clusterv1.MachineDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "progress-test",
+		},
+		Spec: clusterv1.MachineDeploymentSpec{
+			ProgressDeadlineSeconds: pds,
+			Replicas:                &replicas,
+			Strategy: &clusterv1.MachineDeploymentStrategy{
+				Type: clusterv1.RollingUpdateMachineDeploymentStrategyType,
+				RollingUpdate: &clusterv1.MachineRollingUpdateDeployment{
+					MaxUnavailable: intOrStrPtr(0),
+					MaxSurge:       intOrStrPtr(1),
+					DeletePolicy:   pointer.StringPtr("Oldest"),
+				},
+			},
+		},
+		Status: newDeploymentStatus(statusReplicas, updatedReplicas, availableReplicas),
+	}
+	d.Status.Conditions = conditions
+	return d
+}
+
+// helper to create MS with given availableReplicas.
+func newMSWithAvailable(name string, specReplicas, statusReplicas, availableReplicas int32) *clusterv1.MachineSet {
+	return &clusterv1.MachineSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:              name,
+			CreationTimestamp: metav1.Time{},
+			Namespace:         metav1.NamespaceDefault,
+		},
+		Spec: clusterv1.MachineSetSpec{
+			Replicas: pointer.Int32Ptr(specReplicas),
+		},
+		Status: clusterv1.MachineSetStatus{
+			Selector:          "",
+			AvailableReplicas: availableReplicas,
+			Replicas:          statusReplicas,
+		},
+	}
+}
+
+func TestMachineDeploymentSyncDeploymentStatus(t *testing.T) {
+	pds := int32(60)
+	tests := []struct {
+		name               string
+		d                  *clusterv1.MachineDeployment
+		allMSs             []*clusterv1.MachineSet
+		newMS              *clusterv1.MachineSet
+		expectedConditions []*clusterv1.Condition
+	}{
+		{
+			name:   "Deployment not available: MachineDeploymentAvailableCondition should exist and be false",
+			d:      currentDeployment(&pds, 3, 2, 2, 2, clusterv1.Conditions{}),
+			allMSs: []*clusterv1.MachineSet{},
+			newMS:  newMSWithAvailable("foo", 3, 2, 2),
+			expectedConditions: []*clusterv1.Condition{
+				&clusterv1.Condition{
+					Type:     clusterv1.MachineDeploymentAvailableCondition,
+					Status:   corev1.ConditionFalse,
+					Severity: clusterv1.ConditionSeverityWarning,
+					Reason:   clusterv1.MinimumMachinesAvailableReason,
+				},
+			},
+		},
+		{
+			name:   "Deployment Available: MachineDeploymentAvailableCondition should exist and be true",
+			d:      currentDeployment(&pds, 3, 3, 3, 3, clusterv1.Conditions{}),
+			allMSs: []*clusterv1.MachineSet{},
+			newMS:  newMSWithAvailable("foo", 3, 3, 3),
+			expectedConditions: []*clusterv1.Condition{
+				&clusterv1.Condition{
```

Comment:
Shouldn't we expect a ready condition as well? https://github.com/kubernetes-sigs/cluster-api/pull/4625/files#diff-34f64847a48c9fc1b3222c9bec6c93f30767a93262215eb4305d78c3b9097c63R158

Reply:
mm just realised that happens at patching time.